NWNXLib/Services/PerObjectStorage/PerObjectStorage.cpp

```diff
@@ -0,0 +1,224 @@
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetFloatMap().emplace(key, value);
+}
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, std::string value)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetStringMap().emplace(key, value);
+}
+void PerObjectStorage::Set(API::Types::ObjectID object, std::string key, void *value, CleanupFunc cleanup)
+{
+    // TODO: check if the object is actually valid
+    if (object == API::Constants::OBJECT_INVALID)
+        return;
+
+    GetObjectStorage(object).GetPointerMap().emplace(key, std::make_pair<>(value, cleanup));
+}
+
+
+
+void PerObjectStorage::Remove(API::Types::ObjectID object, std::string key)
+{
+    auto it = g_objectStorage.find(object);
+    if (it != g_objectStorage.end())
+    {
+        // Ugly, but GetXxxMap will create it if missing.
+        if (it->second->m_StringMap)
+            it->second->m_StringMap->erase(key);
+        if (it->second->m_IntMap)
+            it->second->m_IntMap->erase(key);
+        if (it->second->m_FloatMap)
+            it->second->m_FloatMap->erase(key);
+        if (it->second->m_PointerMap)
+            it->second->m_PointerMap->erase(key);
+    }
+}
+
+PerObjectStorage::PerObjectStorage()
+{
+
+}
+PerObjectStorage::~PerObjectStorage()
+{
+
+}
+
+PerObjectStorage::ObjectStorage::IntMap& PerObjectStorage::ObjectStorage::GetIntMap()
+{
+    if (!m_IntMap)
+        m_IntMap = std::make_unique<IntMap>();
+    return *m_IntMap;
+}
+PerObjectStorage::ObjectStorage::FloatMap& PerObjectStorage::ObjectStorage::GetFloatMap()
+{
+    if (!m_FloatMap)
+        m_FloatMap = std::make_unique<FloatMap>();
+    return *m_FloatMap;
+}
+PerObjectStorage::ObjectStorage::StringMap& PerObjectStorage::ObjectStorage::GetStringMap()
+{
+    if (!m_StringMap)
+        m_StringMap = std::make_unique<StringMap>();
+    return *m_StringMap;
+}
+PerObjectStorage::ObjectStorage::PointerMap& PerObjectStorage::ObjectStorage::GetPointerMap()
+{
+    if (!m_PointerMap)
+        m_PointerMap = std::make_unique<PointerMap>();
+    return *m_PointerMap;
+}
+
+PerObjectStorage::ObjectStorage::ObjectStorage(API::Types::ObjectID owner)
+{
+    m_oidOwner = owner;
+}
+PerObjectStorage::ObjectStorage::~ObjectStorage()
+{
+    if (m_PointerMap)
+    {
+        for (auto it: *m_PointerMap)
+        {
+            auto ptr = it.second.first;
+            auto cleanup = it.second.second;
+            cleanup(ptr);
+        }
+    }
+}
+
+
+
+PerObjectStorageProxy::PerObjectStorageProxy(PerObjectStorage& perObjectStorage, std::string pluginName)
+    : ServiceProxy<PerObjectStorage>(perObjectStorage)
+{
+    m_pluginName = pluginName;
+}
+PerObjectStorageProxy::~PerObjectStorageProxy()
+{
+    // TODO cleanup all storage from this plugin
```

Comment:
I guess we want each plugin to have its own storage and can't query another plugin's storage. We'll want to give the PerObjectStorageProxy the plugin ID (0 to n) and have the NWNX pointer in CGameObject contain a vector of ObjectStorage or a map of pluginid->storage, one per plugin.

Reply:
Actually, scratch that, I just saw that you're adding the plugin name to each key. That works too.